### DISCUSSION TITLE
Why "next build" does not create all production node_modules?

### DISCUSSION THREAD

**Discussion by ced-mos** [vercel/next.js#46298] [Help]
Hi all,

for our next.js application we want to dockerize the application, where at each startup of the container the commands should be executed to seed and migrate the DB. The commands are provided by the `prisma` package, which is encapsulating the DB. Hence, that this can happens, in the docker container also the `prisma` package need to be existing in the `nodes_modules`.

However, when executing `next build` to create the production build the `prisma` package is not stored into `.next/standalone/node_package`.  This although the package is stored in the `package.json` dependencies and not as a dev dependency.

Only the following packages can be found in the `next/standalone/node_package` folder:
```
   @aws-sdk  cors             @hubspot  next  pino-sentry  @prisma  stripe
..  axios     graphql-request  micro     pino  .pnpm        react    zod
```
Meanwhile I haven't the rule on how `next build` is deciding, if a module is needed or not.

**Therefore, does someone has an idea why this happens and how it can be changed?**

The `next.config.js` used in `next build` looks as follows, where I haven't found an option to change this behavior:
```
/** @type {import('next').NextConfig} */

const nextConfig = {
  reactStrictMode: true,
  swcMinify: true,
  output: 'standalone',
};

module.exports = nextConfig;
```

The `package.json` of my module is as follows:
```
{
  "name": "funkyApp",
  "version": "0.1.0",
  "private": true,
  "scripts": {
    "dev": "next dev",
    "build": "next build",
    "start": "next start",
    "lint": "next lint"
  },
  "prisma": {
    "seed": "ts-node --compiler-options {\"module\":\"CommonJS\"} prisma/seed.ts"
  },
  "dependencies": {
    "@aws-sdk/client-ses": "^3.252.0",
    "@hubspot/api-client": "^8.4.2",
    "@prisma/client": "^4.10.1",
    "@sentry/node": "^7.31.0",
    "@types/react": "18.0.26",
    "@types/react-dom": "18.0.9",
    "ajv": "^8.12.0",
    "axios": "^1.2.2",
    "cors": "^2.8.5",
    "graphql": "^16.6.0",
    "graphql-request": "^5.1.0",
    "micro": "^10.0.1",
    "next": "13.0.6",
    "pino": "^8.8.0",
    "pino-sentry": "^0.14.0",
    "prisma": "^4.8.1",
    "react": "18.2.0",
    "react-dom": "18.2.0",
    "stripe": "^11.6.0",
    "zod": "^3.20.6"
  },
  "devDependencies": {
    "@types/cors": "^2.8.13",
    "@types/node": "18.11.15",
    "eslint": "^8.34.0",
    "eslint-config-next": "^13.1.6",
    "ts-node": "^10.9.1",
    "typescript": "4.9.4"
  }
}
```

Already many thanks for your help.

**Comment by Jahrhause** [comment]
Hi @ced-mos 

Im also dealing with this issue. 

It seems that Next.js in standalone mode, automatically traces all used node-modules an include these in standalone: https://nextjs.org/docs/advanced-features/output-file-tracing#caveats

So it does'nt look in you package.json but only in the code for imports, and those node_modules that it finds here, will be included in the standalone, nodue_modules.

I have sone logging tools wich is not included in the code put as a node scritpt, and this node_module is not included and cant figure out how to included it in next.js

I think that I can manually be copied using Docker script maybe.

It could be nice if these could be included in next.config.js maybe.

I Have tried, using:
```
experimental: {
    outputFileTracingIncludes: {
      '/api/another': ['./necessary-folder/**/*'],
    },
  },
```

but it wont work, as Next.js 13 says:

warn  - Invalid next.config.js options detected: 
  - The value at .experimental has an unexpected property, outputFileTracingIncludes, which is not in the list of allowed properties

**Reply by ced-mos** [reply]
Hi @Jahrhause,

thank you for you message. I tried the `outputFileTracing` feature as well but was not able to achieve the import of my needed module.

For the time being my solution was copying in the Dockerfile the node_modules from the standard `pnpm install` step. However, this leads to a much larger `node_modules` folder, as there a more imported modules but also the modules are much larger as I guess there is no proper minifying step included yet.

Meanwhile, to make sure that the `outputFileTracing` feature is really the thing we would need, can someone from the official maintainer team share some insights, if this feature is, what we need?

**Comment by Jahrhause** [comment]
Hi @ced-mos 

I found a bit of a fix, not pretty but i works.

I one of your routes ex. page.tsx or something like that export a config object with the node modules, you need to include in your standalone folder.

```
import type { PageConfig } from 'next';
...
export const config: PageConfig = {
  unstable_includeFiles: [
    'node_modules/elastic-apm-node/**/*.+(js|json)',
    'node_modules/dotenv/**/*.+(js|json)',
  ],
};
```

credits go to sbstn: https://forum.newrelic.com/s/hubtopic/aAX8W0000008d6sWAA/node-nextjs-with-newrelicnext-and-outputstandalone

**Comment by ced-mos** [comment]
Thank you @Jahrhause for this idea. Need to test it as soon as I will find time.
